crates/matrix-sdk/src/widget/README.md

```diff
@@ -0,0 +1,235 @@
+
+The api follows a request/response pattern. So each request will be sent in both direction.
+This makes for four valid postmessages structures:
+
+- Widget -> Client: `api: FromWidget`, ~~`response:`~~
+- Client -> Widget: `api: FromWidget`, `response: {...}`
+- Client -> Widget: `api: ToWidget`, ~~`response:`~~
+- Widget -> Client: `api: ToWidget`, `response: {...}`
+
+Both APIs conform to the same basic structure:
+
+An example request / response:
+
+```javascript
+{
+    api: $API_NAME,
+    requestId: $REQUEST_ID,
+    action: $ACTION,
+    widgetId: $WIDGET_ID,
+    data: {}
+}
+```
+
+The complete request object is returned to the caller with an additional `response` key like so:
+
+```json
+{
+    api: $API_NAME,
+    requestId: $REQUEST_ID,
+    action: $ACTION,
+    widgetId: $WIDGET_ID,
+    data: {},
+    response: { ... }
+}
+```
+
+The `api` field is required on all requests and must be either `"fromWidget"` or `"toWidget"` depending upon the direction
+of messaging.
+
+The "requestId" field should be unique and included in all requests. This field has been omitted from all of the
+following examples, for the sake of brevity.
+
+The "action" determines the format of the request and response. All actions can return an error response.
+
+The "widgetId" field denotes the widget that the request originates from and is required.
+
+Additional data can be sent as arbitrary fields. However, typically the "data" object should be used.
+
+A success response is an object with zero or more keys where none of keys are “error”.
+
+An error response is a "response" object which consists of a sole "error" key to indicate an error.
+
+They look like:
+
+```json
+{
+    ...
+    error: {
+        message: "Unable to invite user into room.",
+        <Original Error Object>
+    }
+}
+```
+
+The "message" key should be a human-friendly string.
+
+## `WidgetDriver` internals
+
+Internally the driver consists of two main message types:
+
+- [`machine::IncomingMessage`] collects all possible "input" information which can come from the [`MatrixDriver`]
+  or the [`WidgetDriverHandle`] channel, that passes over the actions from the widget.
+  - `WidgetMessage`: An incoming raw message from the widget.
+  - `MatrixDriverResponse`: A response to a request from the `WidgetMachine` to the `MatrixDriver`.
+    For example if the `WidgetMachine` requests the `MatrixDriver` to get approved capabilities,
+    the response would contain what capabilities actually were approved.
+  - `MatrixEventReceived`: The `MatrixDriver` notified the `WidgetMachine` of a new matrix event.
+- [`machine::Action`] describes an operation that the `WidgetDriver` wants to perform:
+  - `SendToWidget`: Send a raw message to the widget.
+  - `MatrixDriverRequest`: A command sent from the client widget API state machine to the
+    client (driver). Once the command is executed,
+    the client will typically generate an `Event` with the result.
```

Comment:
Oh no, another thing we're calling `Event`?

Reply:
I think this mention of `Event` is outdated, I believe it should read `IncomingMessage::MatrixDriverResponse` instead. What's telling is that the docs for `IncomingMessage` say "Incoming **event** that the client API must process" (which, tbh, would benefit from avoiding the term 'event' as well)